Default link colors lack sufficient contrast in the dark color scheme

pxlcoder opened the pull request:
#### 5413aa7d4157f6ff5865f169742255366e984c60
<pre>
Default link colors lack sufficient contrast in the dark color scheme
<a href="https://bugs.webkit.org/show_bug.cgi?id=209851">https://bugs.webkit.org/show_bug.cgi?id=209851</a>
rdar://61149466

Reviewed by Tim Nguyen and Megan Gardner.

Currently, link colors are the same in both the light and dark color schemes.
This results in a lack of contrast against dark backgrounds.

To fix, dark-mode specific link colors are defined. The definitions match the
values used by Chromium, and are not taken from the system link color. This
approach was chosen as the default link color in the light color scheme is
defined by <a href="https://html.spec.whatwg.org/multipage/rendering.html#phrasing-content-3">https://html.spec.whatwg.org/multipage/rendering.html#phrasing-content-3</a>,
and also does not come from the system.

* LayoutTests/css-dark-mode/link-colors-expected.html: Added.
* LayoutTests/css-dark-mode/link-colors.html: Added.
* Source/WebCore/dom/Document.cpp:

Link colors can be adjusted by HTML using the `link`, `alink`, and `vlink`
attributes. Adjust logic to ensure that a static link color is only used if the
corresponding attribute is specified. Without these changes, the dark mode
color would never be used, as the light mode color is stored as the default.

(WebCore::Document::linkColor const):
(WebCore::Document::visitedLinkColor const):
(WebCore::Document::activeLinkColor const):
(WebCore::Document::resetLinkColor):
(WebCore::Document::resetVisitedLinkColor):
(WebCore::Document::resetActiveLinkColor):
* Source/WebCore/dom/Document.h:
(WebCore::Document::linkColor const): Deleted.
(WebCore::Document::visitedLinkColor const): Deleted.
(WebCore::Document::activeLinkColor const): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):

Add definitions for colors under the dark color scheme.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
(WebCore::Style::colorFromPrimitiveValue):

Pass in `StyleColorOptions` when determining the link color, so that it can be
resolved against the used color scheme.

Canonical link: <a href="https://commits.webkit.org/267847@main">https://commits.webkit.org/267847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bfd7c04f3b41e3189ad633ab24933a3b78f32d86

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17779 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18647 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19604 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16630 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17974 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18676 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17992 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18279 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20470 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15507 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22764 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16523 "Passed tests") | [⏳ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20629 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14342 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16048 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16043 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4252 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20410 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16784 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/css-dark-mode/link-colors-expected.html:
```diff
@@ -0,0 +1,16 @@
+</style>
+
+<a class="light" href="https://www.webkit.org">Light</a>
+<a class="dark" href="https://www.webkit.org">Dark</a>
```
You can also add tests for :visited by doing `<a href="">`

pxlcoder replied:
Updated – thanks!

pxlcoder committed:
Default link colors lack sufficient contrast in the dark color scheme
https://bugs.webkit.org/show_bug.cgi?id=209851
rdar://61149466

Reviewed by Tim Nguyen and Megan Gardner.

Currently, link colors are the same in both the light and dark color schemes.
This results in a lack of contrast against dark backgrounds.

To fix, dark-mode specific link colors are defined. The definitions match the
values used by Chromium, and are not taken from the system link color. This
approach was chosen as the default link color in the light color scheme is
defined by https://html.spec.whatwg.org/multipage/rendering.html#phrasing-content-3,
and also does not come from the system.

* LayoutTests/css-dark-mode/link-colors-expected.html: Added.
* LayoutTests/css-dark-mode/link-colors.html: Added.
* Source/WebCore/dom/Document.cpp:

Link colors can be adjusted by HTML using the `link`, `alink`, and `vlink`
attributes. Adjust logic to ensure that a static link color is only used if the
corresponding attribute is specified. Without these changes, the dark mode
color would never be used, as the light mode color is stored as the default.

(WebCore::Document::linkColor const):
(WebCore::Document::visitedLinkColor const):
(WebCore::Document::activeLinkColor const):
(WebCore::Document::resetLinkColor):
(WebCore::Document::resetVisitedLinkColor):
(WebCore::Document::resetActiveLinkColor):
* Source/WebCore/dom/Document.h:
(WebCore::Document::linkColor const): Deleted.
(WebCore::Document::visitedLinkColor const): Deleted.
(WebCore::Document::activeLinkColor const): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):

Add definitions for colors under the dark color scheme.

* Source/WebCore/style/ColorFromPrimitiveValue.cpp:
(WebCore::Style::colorFromPrimitiveValue):

Pass in `StyleColorOptions` when determining the link color, so that it can be
resolved against the used color scheme.

Canonical link: https://commits.webkit.org/267847@main
